lib/puppet/face/facts.rb

```diff
@@ -151,17 +151,19 @@
       options[:resolve_options] = true
       result = Puppet::Node::Facts.indirection.find(Puppet.settings[:certname], options)
 
-      facts = result.values
-
       if options[:value_only]
-        facts.values.first
+        result.values.values.first
       else
-        facts
+        result.values
       end
     end
 
     when_rendering :console do |result|
-      Puppet::Util::Json.dump(result, :pretty => true)
+      if result.is_a?(String)
+        result
+      else
+        Puppet::Util::Json.dump(result, :pretty => true)
+      end
```

Comment:
It looks like facter can emit `[Integer, Float, TrueClass, FalseClass, NilClass, Symbol, String, Array, Hash]`, so I'm thinking we only want to call `Json.dump` for `Array` and `Hash`, and rely on `to_s` for the other types?

Reply:
@joshcooper I think for `NilClass` `JSON.dump` and `to_s` behave differently
```
JSON.dump(nil) #=> "null"
nil.to_s   #=> ""
```

But i'm not sure if Facter will return any `nil` value, or it will convert it before.

Reply:
Facter can return `nil`, so I updated this (including other types in https://github.com/puppetlabs/facter/blob/0654f8f6858ca0163680b3d8dfc53d90345f0f35/lib/facter/custom_facts/util/normalization.rb#L8)

I filed https://tickets.puppetlabs.com/browse/FACT-2930 as it seems facter can emit a `Date` (which seems like a mistake), but the error message about `VALID_TYPES` is inconsistent.